test/multi_instance_test.go

```diff
@@ -12,56 +12,39 @@ import (
 const instanceCount = 4000
 
 func TestMultiSingleton(t *testing.T) {
-	tsg := sim.NewTipSetGenerator(tipSetGeneratorSeed)
-	baseChain := generateECChain(t, tsg)
 	sm, err := sim.NewSimulation(syncOptions(
-		sim.WithHonestParticipantCount(1),
-		sim.WithTipSetGenerator(tsg),
-		sim.WithBaseChain(&baseChain))...)
+		sim.AddHonestParticipant(1, sim.NewUniformECChainGenerator(tipSetGeneratorSeed, 1, 10)),
+	)...)
 	require.NoError(t, err)
-	a := baseChain.Extend(tsg.Sample())
-	sm.SetChains(sim.ChainCount{Count: 1, Chain: a})
-
 	require.NoErrorf(t, sm.Run(instanceCount, maxRounds), "%s", sm.Describe())
-	instance := sm.GetInstance(instanceCount)
+	instance := sm.GetInstance(instanceCount - 1)
```

Comment:
> Does running instanceCount instances no longer leave the sim at instance instanceCount?

Somewhat. What's changed in the simulation logic is what we mean by "leave the sim at".

The changes made in this PR means the simulation will perform the `instanceCount` number of instances, _and_ it does not instantiate any instance that are not going to be run. 

Before the changes made here my understanding was that the fact that the next instance is instantiated but not run is accidental, and that the tests intend to assert the base of latest executed instance instead of the base of next unexecuted instance since to me that seemed more intuitive. Based on the comment above it looks like my understanding was not what the simulation/testing intended to do. 

I will update the code such that it continues to instantiate the next instance after the max number of instances is performed and restore assertions here. 

I don't think "last" and "current" terminology would be clear, since to me it seems strange that the current instance is the one that's beyond the limits given to the simulation (i.e. max instance count). I'll take a shot at introducing reciever functions with clearer names and would love your feedback.

Reply:
> Does running instanceCount instances no longer leave the sim at instance instanceCount?

Somewhat. What's changed in the simulation logic is what the code meant by "leave the sim at".

The changes made in this PR means the simulation will perform at most `instanceCount` number of instances, _and_ it does not instantiate any instance that are not going to be executed. 

Before the changes were mad, my understanding was: the fact that the next instance is instantiated but not executed is accidental, and that the tests intend to assert the base of latest executed instance instead of the base of next unexecuted instance since to me that seemed more intuitive. Based on the comment above it looks like my understanding was not what the simulation/testing intended to do.

The net effect of the changes is the same if the sole reason of using the base of next instance was to get the chain participants agreed on last. Because call to `requireConsensusAtInstance` asserts that consensus matches what it should compared to the instance before last. So I am not entirely sure if there is much more value in making assertions based on the base of next unexecuted instance or the base/head of last 2 executed instance. 

@anorth do you think that's reasonable or we explicitly want to make assertions using the base of next unexecuted instance? If latter, I'd be grateful for your help to understand why.

Reply:
> The changes made in this PR means the simulation will perform at most instanceCount number of instances, and it does not instantiate any instance that are not going to be executed.

Ok

> Before the changes were made, my understanding was: the fact that the next instance is instantiated but not executed is accidental, and that the tests intend to assert the base of latest executed instance instead of the base of next unexecuted instance since to me that seemed more intuitive. Based on the comment above it looks like my understanding was not what the simulation/testing intended to do.

The intention is to assert the _decision_ of the last executed instance.

If that's adequately covered by `requireConsensusAtInstance` then let's delete the additional checks.

Reply:
> I would still expect the final line to be `requireConsensusAtInstance(t, sm, instanceCount-1, expected.Head())` (no `-2`)

I don't follow; why? 

The statement above means you expect the head of last instance's `BaseChain` to be the head of all decisions made at that instance across all participants. Each participant now generates a random chain at each instance, which in this test:
* [is uniform across all of them](https://github.com/filecoin-project/go-f3/pull/187/files#diff-88c39bab87b97682b75decbd3a0a1f8c8b0bb3a5026c740ea86ed6782d76e0f5R16), _and_
* is not the same as instance's `BaseChain`, but [has the same base as  instance's `BaseChain` head](https://github.com/filecoin-project/go-f3/pull/187/files#diff-52bb41335b1f3e988a97b97b651f3f6a683a4641544a86df3aa76c081f0028c1R45-R46).

This means they should all agree on a chain that has the head of instance's `BaseChain` as their decisions _base_, not head. `requireConsensusAtInstance` [makes assertions on decision head](https://github.com/filecoin-project/go-f3/pull/187/files#diff-e5703ff03d7e153511c3591b38829d5757f88469f783418a3c9480d5809b5a9bR42). 

So we either need to:
* change the assertion to use decision's base for your expectation to be true
or
* assert that the head of last instance's `BaseChain` matches the head of decisions made at one-before-last instance (i.e. `instanceCount-2`).

The latter case is what the code currently does.

Does that make sense? What am I missing?

Reply:
> > I would still expect the final line to be `requireConsensusAtInstance(t, sm, instanceCount-1, expected.Head())` (no `-2`)
> 
> I don't follow; why?

Because `expected` should be changed to be the right value/s to expect from the last instance.

I'm trying to make an assertion about the _decision_ made in the last instance. Not about the base chain that went into that instance, not about the base of the decision, but the (head of) the decision itself. I don't believe this is making any assertion about the final decision while a `-2` is involved.

Previously, the expected decision of the last instance was extracted from the base of the _next_ instance, because this could be extracted from the sim. If the sim doesn't roll forward to define that yet, we need another way to extract the expected decision.

Reply:
> > > I would still expect the final line to be `requireConsensusAtInstance(t, sm, instanceCount-1, expected.Head())` (no `-2`)
> > 
> > 
> > I don't follow; why?
> 
> Because `expected` should be changed to be the right value/s to expect from the last instance.

Thank you for your insights, Alex.

The [`expected` variable in the code](https://github.com/filecoin-project/go-f3/blob/4506c1cac54d89396bd64be2f480979e866242de/test/multi_instance_test.go#L20-L22) already represents the last instance completed by the simulation. As pointed out earlier, the change in simulation logic now prevents instantiation beyond the `instanceCount`. Thus, the line `requireConsensusAtInstance(t, sm, instanceCount-1, expected.Head())` should be interpreted as the head of the last instance’s `BaseChain` matching the decision head at that instance, which is an unreasonable expectation in scenarios with EC divergence among participants. I think what the code snippet meant to be is `requireConsensusAtInstance(t, sm, instanceCount-1, <basechain-of-instance-after-last>.Head())`



In terms of the assertion strength, it is neither weaker nor stronger to assert:
  * The head decision at the penultimate instance matches the base of `BaseChain` at the last completed instance, which is what the code currently asserts.
  * The head decision at the last completed instance matches the base of the next instance, which would not be completed due to the `instanceCount` limit, as previously asserted.

Because:

- The simulation [errors if consensus is not reached](https://github.com/filecoin-project/go-f3/blob/4506c1cac54d89396bd64be2f480979e866242de/sim/sim.go#L63-L66) among participants.
- Once consensus is achieved, participants [select the previous round’s decision head as the base for the next chain](https://github.com/filecoin-project/go-f3/blob/4506c1cac54d89396bd64be2f480979e866242de/sim/host.go#L46).
- Increasing the `instanceCount` by one to test a specific instance number achieves the same outcome.

Although we've spent considerable time discussing this, here is how I will proceed:

- I will modify the simulation logic to include an additional instance that will remain incomplete, akin to the previous version.
- I will restore tests to assert the latter scenario since that aligns with your preference, and I have no particular preference myself.

Reply:
@anorth the simulation logic [is now changed](https://github.com/filecoin-project/go-f3/blob/52f910b29c6ea005a77c9f7690894f6b960a0ffe/sim/sim.go#L68-L82) to instantiate `1+instanceCount` where the latest is left incomplete, akin to the previous version.

Please feel free to unresolve this conversation if there is anything else left to do here.